File: backend-py/req/check_duplicate_barcode.py
import requests
from dataclasses import dataclass
from typing import List
from config import STRAPI_URL

@dataclass
class Attributes:
    createdAt: str

@dataclass
class Data:
    id: int
    attributes: Attributes

@dataclass
class Response:
    data: List[Data]


def is_barcode_duplicate(jwt: str, barcode: str, user_id: int) -> bool:
    url = f"{STRAPI_URL}/barcodes?filters[barcode][$eq]={barcode}&filters[users_permissions_user][id][$eq]={user_id}&sort=createdAt:DESC"
    headers = {
        "Authorization": f"Bearer {jwt}"
    }

    response = requests.get(url, headers=headers)
    response.raise_for_status()
    res = response.json()

    response_data = Response(data=[Data(id=item['id'], attributes=Attributes(createdAt=item['attributes']['createdAt'])) for item in res['data']])

    if len(response_data.data) > 0:
        return True
    else:
        return False

File: src-tauri/src/req/check_duplicate_barcode.rs
use serde::Deserialize;
use config::cfg;

#[derive(Deserialize, Debug)]
pub struct Response {
    pub data: Vec<Data>,
}

#[derive(Deserialize, Debug)]
pub struct Data {
    pub id: i32,
    pub attributes: Attributes,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
pub struct Attributes {
    pub createdAt: String,
}

#[tokio::main]
pub async fn is_barcode_duplicate(jwt: &str, barcode: &str, user_id: &i32) -> Result<bool, reqwest::Error> {
    let cfg = cfg();
    let url = format!("{}/api/barcodes?filters[barcode][$eq]={}&filters[users_permissions_user][id][$eq]={}&sort=createdAt:DESC", cfg.api.strapi, barcode, user_id);
    let client = reqwest::Client::new();

    let res = client
        .get(&url)
        .header("Authorization", format!("Bearer {}", jwt))
        .send()
        .await?
        .json::<Response>()
        .await?;

    if res.data.len() > 0 {
        return Ok(true);

    } else {
        return Ok(false);
    }
}